src/dmd/parse.d

```diff
@@ -247,9 +248,23 @@ final class Parser(AST) : Lexer
             /* Create a pseudo-filename for the mixin string, as it may not even exist
              * in the source file.
              */
-            char* filename = cast(char*)mem.xmalloc(strlen(loc.filename) + 7 + (loc.linnum).sizeof * 3 + 1);
-            sprintf(filename, "%s-mixin-%d", loc.filename, cast(int)loc.linnum);
-            scanloc.filename = filename;
+            auto len = strlen(loc.filename) + 7 + (loc.linnum).sizeof * 3 + 1;
+            char[] filename = cast(char[])mem.xmalloc(len)[0 .. len];
+            sprintf(filename.ptr, "%s-mixin-%d", loc.filename, cast(int)loc.linnum);
+
+            // If we are debugging mixins, write `input` out to a file so that the debugger has some
+            // source to look at.
+
+            if (global.params.debugMixins)
+            {
+                auto tmp = FileName.combine(global.params.objdir,filename.ptr);
+                filename = cast(char[])tmp[0 .. strlen(tmp)];
+                auto f = File.create(filename.ptr);
```

Comment:
This is certainly better than the status quo.

However, consider the following code:

void foo(string s)() { mixin(s); }
void main()
{
  foo!(`writeln(a);`);
  foo!(`writeln("hello world");`);
}

This will write `writeln(a);` to file `scratch.d-mixin-5`, then write `writeln("hello world");` to the same file, overwriting what was there previously.

Reply:
Is that likely to ever occur in real code?

Reply:
Any mixin that occurs inside a template that's instantiated more than once. That might be a minority of cases, but it certainly does occur. For instance, std.typecons has at least a handful of these. I wrote some code to automatically extract a postgresql row into a struct, and it contains mixins inside templates that are instantiated multiple times.

Also the standard shortcut for implementing opBinary is with mixins. That's why we switched to opBinary!operator, instead of opAdd and friends.

Reply:
> Is that likely to ever occur in real code?

I think it happens more often than not.
In my experience, most string mixins are wrapped in a mixin template for sanitary reasons, or in a template function. If the enclosing scope is not a template, then there's little motivation for a string mixin... you could just put the code directly. I reckon it's the common case for a mixin to appear in a parametric environment.